Support type-dependent search result highlighting via CSS

timhoffm opened the pull request:
*Edit:* **Summary** (*Update: 25.07.2024*)

This adds result type-dependent HTML classes to the `<li>` elements of the search result list. Implemented types / classes are:

- `context-index`:  If the search term is found in an index such as the glossary
- `context-object`: If the search term a code object such as a python function
- `context-text`: If the search term is found in plain documentation text
- `context-title`: If the search term is found in a heading

The basic style intentionally does not apply styling to these classes to not interfere with derived themes. We encourage theme authors to apply suitable styling to the results list via CSS as they see fit. Check `sphinx13.css` in this PR as an example.

---

It's helpful to visually distinguish different content types.

This PR adds `context` to the javascript result entries (one of "title", "index", "object", "text" - please check that these names make sense, I've inferred them from the JS context) and adds them as a class to the `<li>` item in the result list. This allows styling via CSS.

For simplicity, I've styled with unicode symbols, which should give a decent look without the need to ship our own symbols. Derived themes have the ability to adapt this via CSS and use their own icons if they want to.

Before / after:

![image](https://github.com/sphinx-doc/sphinx/assets/2836374/35f76139-0df4-485f-b3ef-47d393d05998)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ![image](https://github.com/sphinx-doc/sphinx/assets/2836374/9f9a03de-76f7-4a65-b939-5d61ab72b9bb)

jayaddison commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -336,6 +337,7 @@ const Search = {
             null,
             score,
             filenames[file],
+            "title",
```
Perhaps these constant labels could be declared on a JavaScript class, similar to the way we declare some constants on the `Scorer` class in the same `searchtools.js` file?

jayaddison commented on sphinx/themes/basic/static/basic.css_t:
```diff
@@ -114,16 +114,25 @@ img {
-    background-image: url(file.png);
-    background-repeat: no-repeat;
-    background-position: 0 7px;
+    list-style: "\25A1";  /* Unicode: White Square */
```
I wonder whether it might make sense to be slightly more explicit and define these using the CSS `list-style-type` property?

Also, to check: is the white-square default here intended to highlight cases where sites have upgraded their Sphinx CSS but not rebuilt their `searchindex.js`?

timhoffm replied:
I'm not a CSS expert. I was anticipating that `list-style` is easier to override for downstream themes, e.g. `list-style: url(image.svg)`, but just a guess. If you say `list-style-type` is better, I'm happy to change.

I've added the empty square as a catch-all fallback; e.g. could be that somebody rewrites search and includes a new type, but forgets the corresponding CSS. Likely other things can happen and the empty square as as placeholder is better than nothing.

jayaddison replied:
> I'm not a CSS expert. I was anticipating that `list-style` is easier to override for downstream themes, e.g. `list-style: url(image.svg)`, but just a guess. If you say `list-style-type` is better, I'm happy to change.

Me neither, and your intuition may be correct.  I based my suggestion on a sense that `list-style-type` is slightly less ambiguous (`list-style` _does_ accept the same values, but can also be set to an image URL), and also a small note on the [Mozilla MDN page](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style):

> > The `list-style` property is specified as one, two, or three values in any order. If [`list-style-type`](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type) and [`list-style-image`](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-image) are both set, the `list-style-type` is used as a fallback if the image is unavailable.
 
> I've added the empty square as a catch-all fallback; e.g. could be that somebody rewrites search and includes a new type, but forgets the corresponding CSS. Likely other things can happen and the empty square as as placeholder is better than nothing.

That makes sense.  A concern I have with it is that the same (or a similar) symbol is used on some systems when a unicode codepage/symbol is not found in the relevant font, and that could make problems slightly trickier to track down.  Perhaps we could set the value to `initial` (initial browser default)?

AA-Turner replied:
>  A concern I have with it is that the same (or a similar) symbol is used on some systems when a unicode codepage/symbol is not found in the relevant font

Our assumptions in the HTML search is that the user has a modern browser, and hence we should be confident in using Unicode -- every modern browser is able to display emoji, etc.

jayaddison commented on sphinx/themes/basic/static/basic.css_t:
```diff
@@ -114,16 +114,25 @@ img {
+
+ul.search li.title {
+    list-style: "\1F4C1";  /* Unicode: File Folder */
+}
```
This is the only annotation symbol that I find slightly confusing / out of context, personally.

It does sorta logically fit that it's for use to represent that titles are 'containers' in some sense for text content.

I'm browsing unicode character references to try to find something else that could represent text/documents, headings/titles, and/or the potential for containership/nesting.

Edit: remove repetitive phrasing

jayaddison replied:
Maybe it's a controversial misuse, but perhaps the paragraph symbol (called [Pilcrow](https://en.wikipedia.org/wiki/Pilcrow), as I learned today)?

jayaddison replied:
(or we could try to bring the [capitulum](https://en.wikipedia.org/wiki/Capitulum) back into style)

jayaddison replied:
Another alternative could simply be to style `title` and `text` matches identically to begin with.

I'll try to spend a bit more time thinking about this before adding any further thoughts/comments.

timhoffm replied:
The idea behind folder was "structuring element" but I'm not too convinced of it either.

jayaddison replied:
On reflection: I'd recommend keeping this simple, and using the same styling for both `title` and `text` search results - that is, `1F4C4` (unicode: page facing).

We could still _emit_ the CSS classes `title` and `text` in the result list, allowing downstream sites to customize their display to suit their projects if they choose.

Some reasoning for this suggestion: with a default Sphinx configuration and HTML build, it is possible to infer from each listed search result whether the input search query term(s) matched in the corresponding document's title and/or content body.  In my personal opinion it is a nice addition to be able to distinguish between glossary items and programming objects, but seems lower priority to distinguish page title matches from other page content matches.

jayaddison commented on sphinx/themes/basic/static/basic.css_t:
```diff
@@ -115,16 +115,12 @@ img {
-    background-position: 0 7px;
-}
+    padding: 5px 0;
+
```
Missing closing-brace here?

jayaddison reviewed (approved):
Looks good to me - thanks @timhoffm!

![image](https://github.com/sphinx-doc/sphinx/assets/55152140/d073d92e-f7a3-41e2-8511-754e706d8a91)

wlach commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -64,9 +71,10 @@ const _displayItem = (item, searchTerms, highlightTerms) => {
+  const [docName, title, anchor, descr, score, _filename, resultType] = item;
 
   let listItem = document.createElement("li");
+  listItem.classList.add(resultType)
```
Could you add a quick comment explaining what this does?

timhoffm replied:
Do you mean a comment here for clarification, or a code comment for documentation?

Starting here: The search results are written to HTML `<li>` elements - in JS here the `listItem` variable. `resultType` is one of the strings "title", "object", "index", "text". This code line adds the result type as class attribute. e.g. resulting in `<li class="title">`, which can then be in a CSS selector for styling.

wlach replied:
Yeah, something like that but I would probably be a little more succinct: `// Add a class representing the item's type: can be used by a theme's CSS selector for styling` or so.

wlach reviewed (approved):
Sorry a bit late looking at this again.

Approving because I wouldn't be sad if this went in as-is. We'll need a committer (maybe @picnixz or @chrisjsewell?) to actually accept + land it though.

One thing I wonder is whether we should add some documentation about this feature, perhaps somewhere like this:

https://www.sphinx-doc.org/en/master/development/html_themes/index.html

We could also link maintainers of other themes to this as a resource, to get them to adopt/use this new data.

jayaddison commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -64,9 +72,13 @@ const _displayItem = (item, searchTerms, highlightTerms) => {
+  // Add a class representing the item's type:
+  // can be used by a theme's CSS selector for styling
+  // See SearchResultKind for the class names.
+  listItem.classList.add(`result-kind-${resultKind}`);
```
@AA-Turner in https://github.com/sphinx-doc/sphinx/pull/12542 we don't seem to add any prefixes to domain/object-type when those are added as CSS classes.  What's the reason for adding a prefix here?

It feels to me like the CSS class names we choose here could be long-lived; this is the only part of the branch as-is that I feel cautious about.

timhoffm replied:
I had originally not included any prefix, because the typical selector will be `ul.search li.[type]` and `ul.search li.object` seemed clear enough. But I don’t have a strong opinion.

jayaddison replied:
Gradually iterating through some thoughts:

 * From the perspective of a theme author, the ability to reference consistent CSS names when styling search results and ToC entries seems like it'd be beneficial.
 * Sphinx-style object qualifiers (`py:module`, `cpp:class`) would conflict with CSS syntax - so `class="py module"` or `class="cpp class"` might be the preferred output here.
 * Building upon those two thoughts, the `kind` value that we're talking about here is useful context -- but it is a different dimension to the Sphinx object qualifier.  The values could conflict for some domains though -- so in fact, a prefix does seem like a good idea.

jayaddison replied:
One more thought: for some reason I don't love the word _kind_ in this context.  It seems like an unnecessarily-different synonym for `type` -- itself a frequently-overused name in software.

I'd like to express something about what the currently-named-`kind` here relates to.  And I think it's something like a _match category_ -- it's _how_ and/or _why_ the search algorithm decided to include an entry (yes, that's generally derived 1-1 from how the search index is structured, but it doesn't necessarily have to be).

jayaddison replied:
Prototyping / pseudogenerating some results:

```
<result class="match-object cpp class">utils::Format</result>
<result class="match-text">Formatting Guidelines</result>
<result class="match-object py function">foo.bar.format</result>
```

Does that seem semantic and themeable?

jayaddison replied:
(I'm _not_ suggesting that we include Sphinx domain and object type yet -- but trying to anticipate that they could be included in future)

timhoffm replied:
`class="py module"` seems a good extension possibility. I’ve first thought about `class="py-module"` because you likely want to be domain-specific and not style all modules the same way across contexts. However, you can easily combine selectors `.py.module`, which is not worse than `.py-module` and leaves more flexibility.

I’m not having a good prefix either. Personal thoughts: `type` is very generic. Same with `kind`, but that’s additionally feels somewhat unconventional. `match` connects too much to the search process and less to the result. Other possibilities `context`, `category`

Not necessarily related but possibly a good way to come up with suitable names(prefix and type/kind): If we’d add a GitHub-like filter mechanism to the search field, what would be intuitive names for our end users? e.g. `some_func category:api` feels better than`some_func kind:object`.

jayaddison replied:
> Not necessarily related but possibly a good way to come up with suitable names(prefix and type/kind): If we’d add a GitHub-like filter mechanism to the search field, what would be intuitive names for our end users? e.g. `some_func category:api` feels better than`some_func kind:object`.

Mm, good question - query language syntax.  The `label:value` syntax does seem well established for various web-based search engines.  For that specific example, I think that `name:some_func` might be simplest - the function is an object entity that is indexed, and those entities tend to have names.  Other entities could be `document`, `section`, `word`... but I'd also be wary of reinventing the `docutils` node hierarchy.

AA-Turner replied:
> It feels to me like the CSS class names we choose here could be long-lived; this is the only part of the branch as-is that I feel cautious about.

> I had originally not included any prefix, because the typical selector will be `ul.search li.[type]` and `ul.search li.object` seemed clear enough.

I added a prefix out of caution -- CSS backwards-compatibility is very annoying, as you've no idea what selectors people can use. Whilst ``ul.search li.[type]`` is *best*, ``li.[type]`` or even ``.[type]`` is *possible*, and hence reusing the class names somewhere else carries a small risk. Namespacing with some prefix reduces that. I haven't merged #12542 as I'm unsure if the benefit is worth the long-term backward compatibilty commitment. On the other hand, I might be overthinking this whole endevaour entirely.

> One more thought: for some reason I don't love the word _kind_ in this context. It seems like an unnecessarily-different synonym for `type` -- itself a frequently-overused name in software.

*type* felt wrong here, though on a 'feelings' level rather than a great deal of rationality. No strong feelings on going back to *type* really.

A

timhoffm replied:
Having pondered the options, I suggest to go with a prefix `category` or `context`. Both seem equally fine for me. I also suggest to change `object` to `api` because not every item is an object.

jayaddison replied:
I'd like to suggest reserving `category` as a potentially-useful name for some other search-related feature (TBD).

Using `context` seems good to me.  "What context is the search result in?"  "It's in the context of a Python class, a C++ struct, ..."

I'm not so keen on `api` though.  Integer constants and string attributes, for example, don't seem like APIs to me.  They are 'domain objects' though, even if they're not what Python terminology refers to as an `object`.  Also, I can theoretically imagine Sphinx domains that aren't programming languages, but do have some kind of multi-level entity structuring.

Edit: add (important) missing negative.

jayaddison replied:
Also a low-priority argument from my own laziness: I'd like to be able to ask people (with any range of knowledge) "what kind of results did you get?" when they mention a problem with search results, without them potentially thinking that I'm referring to the... whatever we're calling it... of the individual results they saw.

jayaddison replied:
> I'm not so keen on `api` though. Integer constants and string attributes, for example, don't seem like APIs to me. They are 'domain objects' though, even if they're not what Python terminology refers to as an `object`. Also, I can theoretically imagine Sphinx domains that aren't programming languages, but do have some kind of multi-level entity structuring.

Bringing that back to reality: the Sphinx [Domains documentation](https://github.com/sphinx-doc/sphinx/blob/de15d61a46daaaa2b0a0e341cdb4e0abe107e012/doc/usage/domains/index.rst) does state that it is intended to model programming language domains, and does not mention other use-cases

The Sphinx [index definition of `object`](https://www.sphinx-doc.org/en/master/glossary.html#term-object) is currently:

> The basic building block of Sphinx documentation. Every “object directive” (e.g. [py:function](https://www.sphinx-doc.org/en/master/usage/domains/python.html#directive-py-function) or [object](https://www.sphinx-doc.org/en/master/usage/domains/standard.html#directive-object)) creates such a block; and most objects can be cross-referenced to.

(sorry for being verbose)

Edit: clarification.

timhoffm committed:
ENH: Add type-dependent class to search result entries

This allows styling different types of results individually.
It's helpful to visually distinguish different content types.

This PR adds `context` to the javascript result entries
(one of "title", "index", "object", "text") and adds
them as a classes context-title, context-index, context-object,
context-text to the <li> item in the result list.
This allows styling via CSS.

The basic theme only contains the mechanism to add the HTML
classes. It does intentionally not do any styling via CSS.
We reserve that freedom to derived themes.

For the internal sphinx13 theme, I've styled with unicode
symbols, which should give a decent look without the need
to ship our own symbols.

Co-authored-by: Adam Turner <9087854+aa-turner@users.noreply.github.com>

jayaddison commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -47,6 +47,14 @@ if (typeof Scorer === "undefined") {
 }
 
+// Global search result kind enum, used by themes to style search results.
+class SearchResultContext {
```
```suggestion
class SearchResultContext {
```

Maybe a nitpick, but: this was a `const` previously, compared to a `class` now - I'm not familiar enough with JavaScript to know the implications - are there any drawbacks to this change?

timhoffm replied:
This was a change made by @AA-Turner (originally in ebbcb364, which I've squashed). I can't tell about the advantages or disadvantages (also a JavaScript noob), but it was intentional, so I assume it is the way to go.

jayaddison replied:
Ok, thanks (I was too focused on the naming discussion to notice that, evidently :)).

I'll do some background reading out of curiosity (I do notice that the `class` keyword doesn't appear elsewhere in the `basic` theme JavaScript code, and that makes me a little cautious).

jayaddison replied:
Well, it was introduced in ES6 (2015) -- and so was `const` that we already use frequently in the `searchtools.js` code -- so there's no compatibility-level concern.  This seems fine to me - I'll re-scan the rest of the changeset again in a few moments.

(I did notice the word 'kind' remains in a comment -- probably worth updating that to 'context' for consistency?)

jayaddison replied:
Perhaps I am overthinking this, but I think that for both `SearchResultContext` and also the already-existing `Scorer` values, we probably really want immutable, enum-like behaviour.

Yes, OK, perhaps some plugins/themes could usefully and benevolently override them from their own JavaScript code -- but I feel that those use cases should be rare.

Again I'm not an expert but from a quick search, I wonder whether `Object.freeze` is worth considering for both of them: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze

timhoffm replied:
Remark: As long as it does not change behavior, we can always improve the JavaScript later. I wouldn't consider the searchtools.js content as public, but YMMV.

AA-Turner replied:
``class X {}`` is the same as ``const X = class {}`` -- it is syntactic sugar. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#defining_classes

jayaddison replied:
> Remark: As long as it does not change behavior, we can always improve the JavaScript later. I wouldn't consider the searchtools.js content as public, but YMMV.

My understanding of JavaScript is fairly outdated, but I had a sense that any code we add at the top-level pretty much does become public, and that making it private and/or immutable requires special care.

Sufficiently-weird use-cases that rely on released functionality can be difficult to distinguish from denials-of-maintainer-time, so I think we should be careful about the features we do make public.

> `class X {}` is the same as `const X = class {}` -- it is syntactic sugar. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#defining_classes

Yep, true - but the change in https://github.com/sphinx-doc/sphinx/commit/ebbcb364cb63737b92f6b76a2fb35334f1df4709 was from `const X = { ... }` (a constant initialized to an object value) to `class X { ... }`.  I don't yet know whether we want/need `class` semantics -- and am (still) wondering whether `Object.freeze` and perhaps [`Symbol`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) (also ref: [this article](https://hacks.mozilla.org/2015/06/es6-in-depth-symbols/)) could make sense - the goal being to achieve something like an immutable, protected enum.

jayaddison commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,60 @@ If your theme package contains two or more themes, please call
 
+Styling with CSS
+----------------
+The **stylesheets** setting allows to add custom CSS files to a theme.
```
```suggestion
.. tip::

  The :confval:`stylesheets` setting can be used to add custom CSS files to a theme.
```

timhoffm replied:
I've chosen to use the text but not put it in an admonition. First, the sentence describes the foundation of styling; I feel tips should be useful additional but not essential pieces of information. Second, I've turned the following paragraph into a `.. caution`. Having a section only consisting of two admonitions feels weird. There should be some "normal" information.

jayaddison commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,60 @@ If your theme package contains two or more themes, please call
+The structure of the HTML elements and their classes are currently not a
+well-defined public API. Please infer them from inspecting the built HTML
+pages. While we cannot guarantee full stability, they tend to be fairly
+stable.
```
I'd either suggestion a `caution` [admonition](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#admonitions-messages-and-warnings) here, or that we be brave and build consensus that we're OK with the classnames we've selected for this.

jayaddison replied:
Nope, sorry - I'd misunderstood this as being specific to search results, when it clearly is not.  That would lead me towards the admonition suggestion.

jayaddison commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,60 @@ If your theme package contains two or more themes, please call
+- ``ul.search li.context-text``:
+  If the search term is found in plain documentation text
+- ``ul.search li.context-title``:
+  If the search term is found in a heading
```
```suggestion
- ``ul.search li.context-index``:
  For results in an index, such as the glossary
- ``ul.search li.context-object``:
  For results in source code, like Python function definitions
- ``ul.search li.context-title``:
  For results found in section headings
- ``ul.search li.context-text``:
  For results found anywhere else in the documentation text
```

jayaddison commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,60 @@ If your theme package contains two or more themes, please call
+- ``ul.search li.context-title``:
+  If the search term is found in a heading
+
+The basic theme is intentional minimal and does not use them for styling.
```
```suggestion
As a base for inheritance by other themes, the ``basic`` theme is intentionally minimal and does not define CSS rules using these.
```

(maybe too waffly, but I felt the additional background could be useful, despite the fact that we mention derived themes next-sentence)

jayaddison commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,60 @@ If your theme package contains two or more themes, please call
+The basic theme is intentional minimal and does not use them for styling.
+Derived themes are encouraged to use these selectors as they see fit.
+For example, the following code would add context-dependent icons to the
+search result list:
```
```suggestion
For example, the following stylesheet adds contextual icons to the
search result list:
```

timhoffm commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,60 @@ If your theme package contains two or more themes, please call
+The structure of the HTML elements and their classes are currently not a
+well-defined public API. Please infer them from inspecting the built HTML
+pages. While we cannot guarantee full stability, they tend to be fairly
+stable.
```
```suggestion
.. caution::

   The structure of the HTML elements and their classes are currently not a
   well-defined public API. Please infer them from inspecting the built HTML
   pages. While we cannot guarantee full stability, they tend to be fairly
   stable.
```

timhoffm committed:
Apply suggestions from code review

Co-authored-by: James Addison <55152140+jayaddison@users.noreply.github.com>

timhoffm commented on doc/development/html_themes/index.rst:
```diff
@@ -221,6 +221,62 @@ If your theme package contains two or more themes, please call
 
+Styling with CSS
+----------------
+The **stylesheets** setting allows to add custom CSS files to a theme.
```
```suggestion
The :confval:`stylesheets` setting can be used to add custom CSS files to a theme.
```

jayaddison replied:
My mistake, sorry - it seems that this isn't a traditional config setting; it's a theme configuration setting, and those appear to be different.  I'll try to figure out whether it _could_ be made reference-able.. but perhaps that should be reverted.

timhoffm committed:
Update doc/development/html_themes/index.rst

timhoffm committed:
Fix line length from code suggestions

timhoffm committed:
Merge branch 'master' into search-icons

AA-Turner committed:
Merge branch 'master' into search-icons

# Conflicts:
#	CHANGES.rst

AA-Turner committed:
refwarning

AA-Turner committed:
test
